test/browser/components.test.js

```diff
@@ -1573,4 +1574,834 @@ describe('Components', () => {
+				child = this;
+				this.state = { tagName: 'p' };
+			}
+			render() {
+				return h(this.state.tagName, {}, 'Hello');
+			}
+		}
+
+		class Parent1 extends Component {
+			render() {
+				parent1 = this;
+				return this.props.children;
+			}
+		}
+
+		function Sibling(props) {
+			sibling = this;
+			return <p />;
+		}
+
+		function Null() {
+			nullInst = this;
+			return null;
+		}
+
+		render((
+			<Parent1>
+				<Null />
+				<Child />
+				<Sibling />
+			</Parent1>
+		), scratch);
+
+		expect(scratch.innerHTML).to.equal('<p>Hello</p><p></p>');
+		expect(nullInst.base).to.equalNode(null);
+		expect(child.base).to.equalNode(scratch.firstChild);
+		expect(sibling.base).to.equalNode(scratch.lastChild);
+		expect(parent1.base).to.equalNode(child.base);
+
+		child.setState({ tagName: 'span' });
+		rerender();
+
+		expect(scratch.innerHTML).to.equal('<span>Hello</span><p></p>');
+		expect(nullInst.base).to.equalNode(null);
+		expect(child.base).to.equalNode(scratch.firstChild);
+		expect(sibling.base).to.equalNode(scratch.lastChild);
+		expect(parent1.base).to.equalNode(child.base);
+	});
+
+	it('should not update parent c.base if child component changes DOM nodes and a parent is not first child component', () => {
+		let parentDom1;
+		let parent1;
+		let parent2;
+		let child;
+		let sibling;
+
+		class Child extends Component {
+			constructor(props, context) {
+				super(props, context);
+				child = this;
+				this.state = { tagName: 'p' };
+			}
+			render() {
+				return h(this.state.tagName, {}, 'Hello');
+			}
+		}
+
+		class Parent1 extends Component {
+			render() {
+				parent1 = this;
+				return this.props.children;
+			}
+		}
+
+		function Parent2(props) {
+			parent2 = this;
+			return props.children;
+		}
+
+		function ParentWithDom(props) {
+			parentDom1 = this;
+			return <div>{props.children}</div>;
+		}
+
+		function Sibling(props) {
+			sibling = this;
+			return <p />;
+		}
+
+		render((
+			<ParentWithDom>
+				<Parent1>
+					<Sibling />
+					<Parent2>
+						<Child />
+					</Parent2>
+				</Parent1>
+			</ParentWithDom>
+		), scratch);
+
+		expect(scratch.innerHTML).to.equal('<div><p></p><p>Hello</p></div>');
+		expect(child.base).to.equalNode(scratch.firstChild.lastChild);
+		expect(parent2.base).to.equalNode(child.base);
+		expect(sibling.base).to.equalNode(scratch.firstChild.firstChild);
+		expect(parent1.base).to.equalNode(sibling.base);
+		expect(parentDom1.base).to.equalNode(scratch.firstChild);
+
+		child.setState({ tagName: 'span' });
+		rerender();
+
+		expect(scratch.innerHTML).to.equal('<div><p></p><span>Hello</span></div>');
+		expect(child.base).to.equalNode(scratch.firstChild.lastChild);
+		expect(parent2.base).to.equalNode(child.base);
+		expect(sibling.base).to.equalNode(scratch.firstChild.firstChild);
+		expect(parent1.base).to.equalNode(sibling.base);
+		expect(parentDom1.base).to.equalNode(scratch.firstChild);
+	});
+
+	it('should update parent c.base if first child becomes null', () => {
```

Comment:
Had a look into this and it seems like sometimes the `_dom` pointer of components is not set properly. When I log the `vnode` in `unmount()` `_dom` is null whereas `c.base` points to the correct DOM node.

Reply:
Yea, when looking at your comment about a missing test scenario, I realized a while bunch of edge cases I didn't handle lol. Still working on it